src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -416,2571 +457,4397 @@ profiler_exception_clause (
+	_ep_rt_dotnet_runtime_profiler_provider = mono_profiler_create (NULL);
+	_ep_rt_dotnet_mono_profiler_provider = mono_profiler_create (NULL);
+
+	char *diag_env = g_getenv("MONO_DIAGNOSTICS");
+	if (diag_env) {
+		int diag_argc = 1;
+		char **diag_argv = g_new (char *, 1);
+		if (diag_argv) {
+			diag_argv [0] = NULL;
+			if (!mono_parse_options_from (diag_env, &diag_argc, &diag_argv)) {
+				for (int i = 0; i < diag_argc; ++i) {
+					if (diag_argv [i]) {
+						if (strncmp (diag_argv [i], "--diagnostic-mono-profiler=", 27) == 0) {
+							if (!parse_mono_profiler_options (diag_argv [i] + 27))
+								mono_trace (G_LOG_LEVEL_ERROR, MONO_TRACE_DIAGNOSTICS, "Failed parsing MONO_DIAGNOSTICS environment variable option: %s", diag_argv [i]);
+						} else if (strncmp (diag_argv [i], "--diagnostic-ports=", 19) == 0) {
+							char *diag_ports_env = g_getenv("DOTNET_DiagnosticPorts");
+							if (diag_ports_env)
+								mono_trace (G_LOG_LEVEL_WARNING, MONO_TRACE_DIAGNOSTICS, "DOTNET_DiagnosticPorts environment variable already set, ignoring --diagnostic-ports used in MONO_DIAGNOSTICS environment variable");
+							else
+								g_setenv ("DOTNET_DiagnosticPorts", diag_argv [i] + 19, TRUE);
+							g_free (diag_ports_env);
+
+						} else {
+							mono_trace (G_LOG_LEVEL_ERROR, MONO_TRACE_DIAGNOSTICS, "Failed parsing MONO_DIAGNOSTICS environment variable, unknown option: %s", diag_argv [i]);
+						}
+
+						g_free (diag_argv [i]);
+						diag_argv [i] = NULL;
+					}
+				}
+
+				g_free (diag_argv);
+			} else {
+				mono_trace (G_LOG_LEVEL_ERROR, MONO_TRACE_DIAGNOSTICS, "Failed parsing MONO_DIAGNOSTICS environment variable");
+			}
+		}
+	}
+	g_free (diag_env);
+}
+
+void
+ep_rt_mono_init (void)
+{
+	mono_native_tls_alloc (&_ep_rt_mono_thread_holder_tls_id, NULL);
+
+	mono_100ns_ticks ();
+	mono_rand_open ();
+	_ep_rt_mono_rand_provider = mono_rand_init (NULL, 0);
+
+	_ep_rt_mono_initialized = TRUE;
+
+	EP_ASSERT (_ep_rt_default_profiler != NULL);
+	EP_ASSERT (_ep_rt_dotnet_runtime_profiler_provider != NULL);
+	EP_ASSERT (_ep_rt_dotnet_mono_profiler_provider != NULL);
+
+	mono_profiler_set_thread_stopped_callback (_ep_rt_default_profiler, profiler_eventpipe_thread_exited);
+
+	MonoMethodSignature *method_signature = mono_metadata_signature_alloc (mono_get_corlib (), 1);
+	if (method_signature) {
+		method_signature->params[0] = m_class_get_byval_arg (mono_get_object_class());
+		method_signature->ret = m_class_get_byval_arg (mono_get_void_class());
+
+		ERROR_DECL (error);
+		MonoClass *runtime_helpers = mono_class_from_name_checked (mono_get_corlib (), "System.Runtime.CompilerServices", "RuntimeHelpers", error);
+		if (is_ok (error) && runtime_helpers) {
+			MonoMethodBuilder *method_builder = mono_mb_new (runtime_helpers, "CompileMethod", MONO_WRAPPER_RUNTIME_INVOKE);
+			if (method_builder) {
+				_ep_rt_mono_runtime_helper_compile_method = mono_mb_create_method (method_builder, method_signature, 1);
```

Comment:
It will only build a MonoMethod instance, but shouldn't compile the method. It should also handle case where it fails, and in that case we won't emit the phantom frame in callstacks.